src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java

```diff
@@ -85,6 +98,188 @@ public Set<StudentDTO> registerMultipleStudents(TutorialGroup tutorialGroup, Set
+        var uniqueRegistrationsWithMatchingUsers = new HashSet<Pair<TutorialGroupRegistrationImportDTO, User>>();
+        for (var registrationAndMatchingUser : registrationsWithMatchingUsers) {
+            var user = registrationAndMatchingUser.getSecond();
+            var registrationsForUser = userToRegistrations.get(user);
+            if (registrationsForUser.size() > 1) {
+                failedRegistrations.add(registrationAndMatchingUser.getFirst().withImportResult(false, TutorialGroupImportErrors.MULTIPLE_REGISTRATIONS));
+            }
+            else {
+                uniqueRegistrationsWithMatchingUsers.add(registrationAndMatchingUser);
+            }
+        }
+        return uniqueRegistrationsWithMatchingUsers;
+    }
+
+    private Set<TutorialGroup> findOrCreateTutorialGroups(Course course, Set<TutorialGroupRegistrationImportDTO> registrations) {
+        var titlesMentionedInRegistrations = registrations.stream().map(TutorialGroupRegistrationImportDTO::title).map(String::trim).collect(Collectors.toSet());
+
+        var foundTutorialGroups = tutorialGroupRepository.findAllByCourseId(course.getId()).stream()
+                .filter(tutorialGroup -> titlesMentionedInRegistrations.contains(tutorialGroup.getTitle())).collect(Collectors.toSet());
+        var tutorialGroupsToCreate = titlesMentionedInRegistrations.stream()
+                .filter(title -> foundTutorialGroups.stream().noneMatch(tutorialGroup -> tutorialGroup.getTitle().equals(title))).map(title -> new TutorialGroup(course, title))
+                .collect(Collectors.toSet());
+
+        var tutorialGroupsMentionedInRegistrations = new HashSet<>(foundTutorialGroups);
+        tutorialGroupsMentionedInRegistrations.addAll(tutorialGroupRepository.saveAll(tutorialGroupsToCreate));
+        return tutorialGroupsMentionedInRegistrations;
+    }
+
+    private Set<TutorialGroupRegistrationImportDTO> filterOutWithoutTitle(Set<TutorialGroupRegistrationImportDTO> registrations,
+            Set<TutorialGroupRegistrationImportDTO> failedRegistrations) {
+        var registrationsWithTitle = new HashSet<TutorialGroupRegistrationImportDTO>();
+        var registrationsWithoutTitle = new HashSet<TutorialGroupRegistrationImportDTO>();
+        for (var importDTO : registrations) {
+            if (importDTO.title() == null || importDTO.title().isBlank()) {
+                registrationsWithoutTitle.add(importDTO.withImportResult(false, TutorialGroupImportErrors.NO_TITLE));
+            }
+            else {
+                registrationsWithTitle.add(importDTO);
+            }
+        }
+        failedRegistrations.addAll(registrationsWithoutTitle);
+        return registrationsWithTitle;
+    }
+
+    private Set<Pair<TutorialGroupRegistrationImportDTO, User>> filterOutWithoutMatchingUser(Course course, Set<TutorialGroupRegistrationImportDTO> registrations,
+            HashSet<TutorialGroupRegistrationImportDTO> failedRegistrations) {
```

Comment:
```suggestion
    private Set<Pair<TutorialGroupRegistrationImportDTO, User>> filterOutWithoutMatchingUser(Course course, Set<TutorialGroupRegistrationImportDTO> registrations,
            Set<TutorialGroupRegistrationImportDTO> failedRegistrations) {
```
Codacy

Reply:
Done in 16c16c0a2c8e3df8e3aa4fff091720438aefa76f